<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <!-- Parameters -->
    <arg name="use_sim_time" default="true"/>
    <param name="use_sim_time" value="$(arg use_sim_time)"/>

    <!-- GMSF Input Topics -->
    <arg name="imu_topic_name" default="/zed2i/zed_node/imu/data"/>
    <arg name="gnss_topic_name" default="/rtk_gps_driver/position_receiver_0/ros/navsatfix"/>
    <arg name="lidar_odometry_topic_name" default="/slam_optimized_poses_odometry"/>
    <arg name="legged_odometry_pose_topic_name" default="/state_estimator/pose_in_odom22"/>
    <arg name="legged_odometry_odom_topic_name" default="/zed2i/zed_node/odom"/>
    <arg name="legged_kinematics_topic_name" default="/state_estimator/anymal_state22"/>


     <arg name="rviz_file_path" default="$(find anymal_estimator_graph)/rviz/gnss_lidar.rviz"/>
    <!-- Servers -->
    <arg name="gnss_coordinates_to_enu_topic_name" default="/dual_graph_node/get_path_in_enu"/>

    <!-- Config -->
    <arg name="extrinsics_from_urdf" default="false"/>
    <arg name="using_gnss_unary" default="false"/>
    <arg name="using_lio_unary" default="true"/>
    <arg name="using_lio_between" default="false"/>
    <arg name="using_legged_between" default="true"/>
    <arg name="using_legged_velocity_unary" default="false"/>
    <arg name="using_legged_kinematics" default="false"/>

    <!-- Config Files -->
    <!-- GNSS Available -->
    <arg if="$(arg using_gnss_unary)" name="core_graph_param_file"
         default="$(find anymal_estimator_graph)/config/core/core_graph_params_gnss.yaml"/>
    <arg if="$(arg using_gnss_unary)" name="anymal_graph_param_file"
         default="$(find anymal_estimator_graph)/config/anymal_specific/anymal_graph_params_gnss.yaml"/>
    <!-- GNSS Unavailable -->
    <arg unless="$(arg using_gnss_unary)" name="core_graph_param_file"
         default="$(find anymal_estimator_graph)/config/core/core_graph_params_no_gnss.yaml"/>
    <arg unless="$(arg using_gnss_unary)" name="anymal_graph_param_file"
         default="$(find anymal_estimator_graph)/config/anymal_specific/anymal_graph_params_no_gnss.yaml"/>
    <!-- GNSS Independent -->
    <arg name="core_graph_config_param_file"
         default="$(find anymal_estimator_graph)/config/core/core_graph_config.yaml"/>
    <arg name="core_extrinsic_param_file"
         default="$(find anymal_estimator_graph)/config/core/core_extrinsic_params.yaml"/>
    <arg name="anymal_extrinsic_param_file"
         default="$(find anymal_estimator_graph)/config/anymal_specific/anymal_extrinsic_params.yaml"/>
    <arg name="anymal_gnss_param_file"
         default="$(find anymal_estimator_graph)/config/anymal_specific/anymal_gnss_params.yaml"/>
    <arg name="anymal_traj_alignment_param_file"
         default="$(find anymal_estimator_graph)/config/anymal_specific/anymal_traj_align_params.yaml"/>

    <!-- Logging Location -->
    <arg name="logging_dir_location" default="$(find anymal_estimator_graph)/logging"/>

    <!-- Node -->
    <node pkg="anymal_estimator_graph" type="anymal_estimator_graph_node" name="anymal_estimator_node"
          output="screen"> <!--launch-prefix="gdb -ex run -args"-->
        <!-- Launch Parameters -->
        <param name="launch/usingGnssUnary" type="bool" value="$(arg using_gnss_unary)"/>
        <param name="launch/usingLioUnary" type="bool" value="$(arg using_lio_unary)"/>
        <param name="launch/usingLioBetween" type="bool" value="$(arg using_lio_between)"/>
        <param name="launch/usingLeggedBetween" type="bool" value="$(arg using_legged_between)"/>
        <param name="launch/usingLeggedVelocityUnary" type="bool" value="$(arg using_legged_velocity_unary)"/>
        <param name="launch/usingLeggedKinematics" type="bool" value="$(arg using_legged_kinematics)"/>
        <param name="launch/optimizationResultLoggingPath" type="str" value="$(arg logging_dir_location)"/>
        <!-- Parameter files -->
        <rosparam command="load" file="$(arg core_graph_config_param_file)"/>
        <rosparam command="load" file="$(arg core_graph_param_file)"/>
        <rosparam command="load" file="$(arg core_extrinsic_param_file)"/>
        <rosparam command="load" file="$(arg anymal_graph_param_file)"/>
        <rosparam command="load" file="$(arg anymal_extrinsic_param_file)"/>
        <rosparam command="load" file="$(arg anymal_gnss_param_file)"/>
        <rosparam command="load" file="$(arg anymal_traj_alignment_param_file)"/>
        <!-- Remapping of topics -->
        <remap from="/imu_topic" to="$(arg imu_topic_name)"/>
        <remap from="/lidar_odometry_topic" to="$(arg lidar_odometry_topic_name)"/>
        <remap from="/gnss_topic" to="$(arg gnss_topic_name)"/>
        <remap from="/legged_odometry_pose_topic" to="$(arg legged_odometry_pose_topic_name)"/>
        <remap from="/legged_odometry_odom_topic" to="$(arg legged_odometry_odom_topic_name)"/>
        <remap from="/anymal_state_topic" to="$(arg legged_kinematics_topic_name)"/>
        <!-- Remapping of services -->
        <remap from="/gnss_coordinates_to_enu_topic" to="$(arg gnss_coordinates_to_enu_topic_name)"/>
    </node>



    <!-- RVIZ for Visualization -->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(arg rviz_file_path)" required="true" output="screen"/>

    <!-- TF for Aligning Compslam Map 
    <node pkg="tf" type="static_transform_publisher" name="compslam_map_to_gmsf_map" args="0 0 0 0 0 0 map_o3d map_graph_msf_ 10" />-->

</launch>
